### ISSUE TITLE
code for quark jet

### ISSUE THREAD

**Issue by avyuktjindal2010-art** [avyuktjindal2010-art/tao-bend-framework#1]
`import os
import math
import numpy as np
import pandas as pd
import torch
import numba
from scipy.optimize import minimize, Bounds
import emcee
from emcee.moves import DEMove, WalkMove, StretchMove
import matplotlib.pyplot as plt

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# -------------------------
# 1) Data loaders
# -------------------------
def load_table(path):
    df = pd.read_csv(path, header=None, comment="#", engine="python", sep=None)
    df = df.apply(pd.to_numeric, errors="coerce")
    df = df[df.notna().any(axis=1)]
    n = df.shape[1]
    if n == 2:
        r, y = df[0].to_numpy(float), df[1].to_numpy(float)
        err = np.ones_like(y)
    elif n == 3:
        r, y, err = df[0].to_numpy(float), df[1].to_numpy(float), df[2].to_numpy(float)
    else:
        r, y = df[0].to_numpy(float), df[1].to_numpy(float)
        stat, syst = df[2].to_numpy(float), df[3].to_numpy(float)
        err = np.sqrt(stat*stat + syst*syst)
    return r, y, err

def load_all_tables(base_dir="."):
    Rs, Ys, Es = [], [], []
    for k in range(1, 10):
        p = os.path.join(base_dir, f"Table{k}.csv")
        if os.path.exists(p):
            r, y, e = load_table(p)
            Rs.append(r); Ys.append(y); Es.append(e)
    if not Rs:
        raise FileNotFoundError("No Table1..9.csv found.")
    return np.concatenate(Rs), np.concatenate(Ys), np.concatenate(Es)

# -------------------------
# 2) Tao–Bend model
# -------------------------
def qleak_gpu(kappa, beta, lamTB, T, mu, bend, resist):
    dtype = torch.complex64
    sx = torch.tensor([[0,1],[1,0]], dtype=dtype, device=device)
    sz = torch.tensor([[1,0],[0,-1]], dtype=dtype, device=device)
    H = (kappa + 0.1*T)*sz + (lamTB + 0.05*mu)*sx
    psi0 = torch.zeros(2, dtype=dtype, device=device)
    psi0[0] = 1+0j
    t = torch.linspace(0, 4, 200, device=device)
    gamma = 0.1*(bend+resist)
    U = torch.matrix_exp((-1j*H).unsqueeze(0)*t[:,None,None])
    psi = U @ psi0
    psi = psi * torch.exp(-gamma*t).unsqueeze(1)
    sxpsi = psi @ sx
    corr = torch.real(torch.sum(torch.conj(psi)*sxpsi, dim=1))
    corr = torch.clamp(corr, 0)
    trap = torch.trapz(corr, t)/4
    return float(torch.clamp(1-trap, 0, 0.98))

@numba.njit(fastmath=True)
def tao_bend_base(r, r2, r4, th):
    kappa, beta, lamTB = th[0], th[1], th[2]
    epsL, lamL         = th[3], th[4]
    s50, s30, s10, s0  = th[5], th[6], th[7], th[8]
    bend, resist       = th[9], th[10]
    spinB              = th[11]
    tau, lr2, lr4      = th[12], th[13], th[14]
    scale = 0.25*(s50 + s30 + s10 + s0)
    base = (1 + kappa*r2 + lamTB*r4)*np.exp(-beta*r2)
    base += tau*(6*r2 - 2)*np.exp(-beta*r2)
    base += bend*resist*(2*r)
    rough = epsL + lr2*r2 + lr4*r4
    return (base*(1-rough))*scale

def tao_bend(r, th):
    r2 = r*r
    r4 = r2*r2
    tb = tao_bend_base(r, r2, r4, th)
    leak = qleak_gpu(th[0], th[1], th[2], 0.25, 0.05, th[9], th[10])
    return tb*(1 - th[3] - th[4]*leak)

# -------------------------
# 3) Noise model (learned u)
# -------------------------
def sigma_phys(err, y, scale, u):
    stat   = scale * err
    add    = 0.10 * np.abs(y)
    unfold = u * np.abs(y)
    floor  = 0.03
    return np.sqrt(stat*stat + add*add + unfold*unfold + floor*floor)

# -------------------------
# 4) Priors (include u and c_corr) with tightened bounds
# -------------------------
def log_prior(th):
    # th length = 18: 0..15 original, 16=u, 17=c_corr
    if not (0 < th[0] < 5): return -np.inf
    if not (0.1 < th[1] < 10): return -np.inf
    if not (0 < th[2] < 3): return -np.inf
    if not (0 <= th[3] < 1): return -np.inf
    if not (0 <= th[4] < 2): return -np.inf
    for v in th[5:9]:
        if not (0.1 < v < 3): return -np.inf
    if not (0.01 < th[9]  < 3): return -np.inf
    if not (0.01 < th[10] < 3): return -np.inf
    if not (0.5  < th[11] < 4): return -np.inf
    if not (0 <= th[12] < 5): return -np.inf
    if not (0 <= th[13] < 5): return -np.inf
    if not (0 <= th[14] < 5): return -np.inf
    # log(s) prior
    if not (-2.5 < th[15] < 1.5): return -np.inf
    # u prior (reduced)
    if not (0.05 <= th[16] <= 0.12): return -np.inf
    # c_corr prior (reduced)
    if not (0.6 <= th[17] <= 1.4): return -np.inf
    return 0.0

# -------------------------
# 5) Rank-8 correlated modes with reduced base amplitudes
# -------------------------
def build_correlated_modes(r, y, c_corr=1.0):
    v1 = y
    v2 = r*y
    v3 = np.power(np.abs(y) + 1e-12, 0.8)
    v4 = np.ones_like(y)
    v5 = (r - np.mean(r)) * y
    v6 = np.log(r + 1e-3) * y
    v7 = (r*r) * y
    v8 = np.sin(np.pi*r) * y

    # reduced base amplitudes (35% lower than previous)
    amps = np.array([0.36, 0.28, 0.21, 0.15, 0.12, 0.12, 0.10, 0.10]) * c_corr
    W = np.column_stack([
        amps[0]*v1,
        amps[1]*v2,
        amps[2]*v3,
        amps[3]*v4,
        amps[4]*v5,
        amps[5]*v6,
        amps[6]*v7,
        amps[7]*v8
    ])
    return W

# -------------------------
# 6) Full likelihood (Woodbury) uses u and c_corr
# -------------------------
def log_like(th, r, y, err):
    ym = tao_bend(r, th)
    s  = math.exp(th[15])
    u  = th[16]
    c_corr = th[17]

    sig = sigma_phys(err, y, s, u)
    var = sig*sig
    Dinv = 1.0 / var
    rvec = y - ym

    W = build_correlated_modes(r, y, c_corr=c_corr)
    WDinv = (W.T * Dinv)
    B = np.eye(W.shape[1]) + WDinv @ W

    signB, logdetB = np.linalg.slogdet(B)
    if signB <= 0:
        return -np.inf

    rD = rvec * Dinv
    rD_W = rD @ W
    B_inv = np.linalg.inv(B)

    chi2_diag = np.dot(rvec, rD)
    chi2_corr = chi2_diag - rD_W @ B_inv @ rD_W
    logdetD = np.sum(np.log(var))

    return -0.5 * (chi2_corr + logdetD + logdetB)

def log_post(th, r, y, err):
    lp = log_prior(th)
    if lp == -np.inf: return -np.inf
    ll = log_like(th, r, y, err)
    if not np.isfinite(ll): return -np.inf
    return lp + ll

# -------------------------
# 7) PT-MCMC (slightly increased nwalk for robustness)
# -------------------------
def run_pt_mcmc(th_map, r, y, err):
    ndim = len(th_map)
    nwalk = 56
    temps = [1.0, 1.75, 3.0, 6.0]
    moves = [DEMove(), WalkMove(), StretchMove(a=2.8)]
    p0 = [th_map + 0.02*np.random.randn(nwalk, ndim) for _ in temps]

    samplers = []
    for i, T in enumerate(temps):
        def logp(theta, TT=T): return log_post(theta, r, y, err) / TT
        sampler = emcee.EnsembleSampler(nwalk, ndim, logp, moves=moves)
        samplers.append(sampler)

    for i, sampler in enumerate(samplers):
        print(f"[PT] Temp {temps[i]}")
        sampler.run_mcmc(p0[i], 700, progress=True)

    chain = samplers[0].get_chain(flat=True)
    acc = np.mean(samplers[0].acceptance_fraction)
    return chain, acc

# -------------------------
# 8) Diagnostics and plotting
# -------------------------
def compute_residuals(r, y, err, th):
    ym = tao_bend(r, th)
    s  = math.exp(th[15])
    u  = th[16]
    sig = sigma_phys(err, y, s, u)
    res = y - ym
    pulls = res / sig
    return res, pulls, sig

def plot_residuals(r, res, pulls):
    plt.figure(figsize=(7,4))
    plt.axhline(0, color='k')
    plt.plot(r, res, 'o', ms=4)
    plt.grid(alpha=0.3)
    plt.title("Residuals")
    plt.tight_layout(); plt.savefig("tb_residuals.png", dpi=200); plt.close()

    plt.figure(figsize=(7,4))
    plt.axhline(0, color='k'); plt.axhline(1, ls='--', color='gray'); plt.axhline(-1, ls='--', color='gray')
    plt.plot(r, pulls, 'o', ms=4)
    plt.grid(alpha=0.3)
    plt.title("Pulls")
    plt.tight_layout(); plt.savefig("tb_pulls.png", dpi=200); plt.close()

# -------------------------
# 9) WAIC (diag approx)
# -------------------------
def log_like_diag_pointwise(th, r, y, err):
    ym = tao_bend(r, th)
    s  = math.exp(th[15])
    u  = th[16]
    sig = sigma_phys(err, y, s, u)
    var = sig*sig
    res2 = (y - ym)**2 / var
    log2pi = math.log(2*math.pi)
    return -0.5*(res2 + np.log(var) + log2pi)

def waic_tb(samples, r, y, err, max_draws=200):
    ns = samples.shape[0]
    draws = min(ns, max_draws)
    idx = np.random.choice(ns, draws, replace=False)
    N = len(r)
    ll = np.zeros((draws, N))
    for i, j in enumerate(idx):
        th = samples[j]
        ll[i] = log_like_diag_pointwise(th, r, y, err)
    lppd = np.sum(np.log(np.mean(np.exp(ll), axis=0)))
    p_waic = np.sum(np.var(ll, axis=0, ddof=1))
    waic = -2*(lppd - p_waic)
    return waic, p_waic

# -------------------------
# 10) PPC on χ²
# -------------------------
def ppc_chi2(samples, r, y, err, n_draws=200):
    ns = samples.shape[0]
    draws = min(ns, n_draws)
    idx = np.random.choice(ns, draws, replace=False)
    chi2_rep = []
    for j in idx:
        th = samples[j]
        ym = tao_bend(r, th)
        s  = math.exp(th[15])
        u  = th[16]
        sig = sigma_phys(err, y, s, u)
        yrep = np.random.normal(loc=ym, scale=sig)
        res = (yrep - ym)/sig
        chi2_rep.append(np.sum(res*res))
    arr = np.array(chi2_rep)
    return {
        "mean": float(arr.mean()),
        "std": float(arr.std()),
        "q05": float(np.quantile(arr, 0.05)),
        "q50": float(np.quantile(arr, 0.50)),
        "q95": float(np.quantile(arr, 0.95)),
    }

# -------------------------
# 11) Main pipeline
# -------------------------
def run_analysis():
    r, y, err = load_all_tables(".")
    print("[INFO] Loaded", len(r), "points")

    # theta has length 18: 0..15 params, 16=u, 17=c_corr
    th0 = np.array([
        1.0, 3.0, 0.6, 0.1, 0.6,
        1.1, 0.9, 0.85, 0.8,
        0.7, 0.8, 2.0,
        0.5, 0.1, 0.05,
        0.0,    # log s
        0.10,   # u initial
        0.80    # c_corr initial (reduced)
    ])

    low = np.array([0,0.1,0,0,0, 0.1,0.1,0.1,0.1, 0.01,0.01,0.5, 0,0,0, -2.5, 0.05, 0.6])
    up  = np.array([5,10,3,1,2, 3,3,3,3, 3,3,4, 5,5,5, 1.5, 0.12, 1.4])

    print("[INFO] MAP optimization ...")
    def neg_log_post(th):
        v = log_post(th, r, y, err)
        return 1e30 if not np.isfinite(v) else -v

    mapres = minimize(neg_log_post, th0, method="L-BFGS-B", bounds=Bounds(low, up), options={"maxiter":3000})
    th_map = mapres.x
    print("[MAP] θ_map =", th_map, "| s_map =", math.exp(th_map[15]), "| u_map =", th_map[16], "| c_map =", th_map[17])

    print("[INFO] PT-MCMC ...")
    samples, acc = run_pt_mcmc(th_map, r, y, err)
    th_best = np.mean(samples, axis=0)
    print("[MCMC] θ_best =", th_best)
    print("[MCMC] s_best =", math.exp(th_best[15]))
    print("[MCMC] u_best =", th_best[16])
    print("[MCMC] c_corr_best =", th_best[17])
    print("[MCMC] acceptance =", acc)

    ym = tao_bend(r, th_best)
    s  = math.exp(th_best[15])
    u  = th_best[16]
    sig = sigma_phys(err, y, s, u)
    chi2 = np.sum(((y - ym)/sig)**2)
    dof = len(r) - len(th_best)
    chi2_red = chi2 / dof
    print("\n===== FINAL RESULTS =====")
    print("χ² =", chi2, "| dof =", dof, "| χ²_red =", chi2_red)

    res, pulls, _ = compute_residuals(r, y, err, th_best)
    plot_residuals(r, res, pulls)
    print("[INFO] Saved residual plots")

    waic, p_waic = waic_tb(samples, r, y, err)
    print("\n[INFO] WAIC (diag approx) =", waic, "| p_WAIC =", p_waic)

    ppc = ppc_chi2(samples, r, y, err)
    print("\n[INFO] PPC χ² summary:", ppc)

    return {
        "theta_best": th_best,
        "chi2_red": chi2_red,
        "s_best": math.exp(th_best[15]),
        "u_best": th_best[16],
        "c_corr_best": th_best[17],
        "waic": waic,
        "p_waic": p_waic,
        "ppc": ppc,
        "acceptance": acc
    }

if __name__ == "__main__":
    out = run_analysis()
    print("\n===== SUMMARY =====")
    for k,v in out.items():
        print(k, ":", v)`
